src/clj/puppetlabs/services/legacy_routes/legacy_routes_core.clj

```diff
@@ -0,0 +1,120 @@
+(ns puppetlabs.services.legacy-routes.legacy-routes-core
+  (:require [clojure.string :as str]
+            [clojure.tools.logging :as log]
+            [compojure.core :as compojure]
+            [ring.util.codec :as ring-codec]
+            [compojure.route :as route]
+            [clojure.tools.logging :as log]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; v3 => v4 header translation support functions
+
+; TODO: Improve the naive parsing here with a real library like
+; https://github.com/ToBeReplaced/http-accept-headers
+(defn- map-accept-header
+  "Convert Accept: raw, s media types to Accept: binary.  NOTE: This method does
+  not conform to RFC-2616 Accept: header specification.
+
+  The split on `,` is naive, the method should take into account `;` parameters
+  and such to be RFC compliant, but puppet agent does not use this format so
+  implementing the full spec is unnecessary."
+  [request]
+  (if-let [accept (get-in request [:headers "accept"])]
+    (let [vals (map str/lower-case (str/split accept #"\s*,\s*"))]
+      (assoc-in request [:headers "accept"]
+        (->> vals
+          (replace {"raw" "binary", "s" "binary"})
+          (distinct)
+          (str/join ", "))))
+    request))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Handler wrapper
+
+(defn request-compatibility-wrapper
+  "Given a Puppet 3.X CA request, convert it to a Puppet 4.X CA request."
+  [handler mount-point api-version request]
+  (let [{{environment :environment} :params
+         path-info                  :path-info
+         query-string               :query-string} request
+        query-params (if query-string (ring-codec/form-decode query-string) {})]
```

Comment:
I ran into a problem with this around one of the CA requests that a Puppet 3.x agent makes.  Here's the original request:

~~~~
GET /production/certificate/agent929?
~~~~

In this case, `query-string` ends up being an empty string, is passed into `form-decode`, is returned back as an empty string, and results in this error being returned to the client:

~~~~
Error: Could not request certificate: Error 500 on SERVER: <html>
<head>
<meta http-equiv="Content-Type" content="text/html;charset=ISO-8859-1"/>
<title>Error 500 </title>
</head>
<body>
<h2>HTTP ERROR: 500</h2>
<p>Problem accessing /production/certificate/agent929. Reason:
<pre>    java.lang.String cannot be cast to clojure.lang.IPersistentCollection</pre></p>
<hr /><i><small>Powered by Jetty://</small></i>
</body>
</html>
~~~~

I was able to work around this by changing this line to:

~~~~clj
query-params (if (str/blank? query-string)
                         {}
                         (ring-codec/form-decode query-string))
~~~~

Not sure if that's the best change but got me past this problem.